src/Raven.Server/Documents/Queries/LuceneIntegration/TermsMatchQuery.cs

```diff
@@ -19,115 +22,191 @@ public TermsMatchQuery(string field, List<string> matches)
             Field = field;
             Matches = matches;
             Matches.Sort(StringComparer.Ordinal);
-            RewriteMethod = CONSTANT_SCORE_FILTER_REWRITE;
         }
 
-        public override string ToString(string fld)
+
+        public override Weight CreateWeight(Searcher searcher, IState state)
         {
-            return "@in<" + Field + ">(" + string.Join(", ", Matches) + ")";
+            return new TermMatchQueryWeight(this, searcher);
         }
 
-        protected override FilteredTermEnum GetEnum(IndexReader reader, IState state)
+        private class TermMatchQueryWeight : Weight
         {
-            return new RavenTermsFilteredTermEnum(this, reader, state);
-        }
+            private readonly TermsMatchQuery _parent;
+            private readonly Searcher _searcher;
+            private float _queryWeight = 1.0f;
+
+            public TermMatchQueryWeight(TermsMatchQuery parent, Searcher searcher)
+            {
+                _parent = parent;
+                _searcher = searcher;
+            }
+            public override Lucene.Net.Search.Explanation Explain(IndexReader reader, int doc, IState state)
+            {
+                var result = new ComplexExplanation {Description = _parent.ToString(), Value = _queryWeight};
+                result.AddDetail(new Lucene.Net.Search.Explanation(_queryWeight, "queryWeight"));
+                return result;
+            }
 
-        private sealed class RavenTermsFilteredTermEnum : FilteredTermEnum
+            public override void Normalize(float norm)
+            {
+                _queryWeight *= norm;
+            }
+
+            public override Scorer Scorer(IndexReader reader, bool scoreDocsInOrder, bool topScorer, IState state)
+            {
+                Similarity similarity = _parent.GetSimilarity(_searcher);
+                if(_parent.Matches.Count > 128)
+                    return new LazyInitTermMatchScorer(_parent, reader, state, similarity);
+
+                var scorers = new Scorer[_parent.Matches.Count];
+                int index = 0;
+                foreach (var match in _parent.Matches)
+                {
+                    var termDocs = reader.TermDocs(new Term(_parent.Field, match),state);
+                    var scorer = new TermScorer(this, termDocs,similarity, reader.Norms(_parent.Field, state));
```

Comment:
Can we call ` reader.Norms(_parent.Field, state)` just once before the loop and pass the returned array here?

Reply:
Yes, there should be no reason to do so.